### PR TITLE
feat(api): Start API server after first L1 batch

### PR DESCRIPTION
**Pull request by slowli**
## What ❔

- Delays the API server start until there's at least one L1 batch present in the node DB.
- Refactors the VM sandbox so that it's more modular / maintainable.
- Tests VM-related API server methods.

## Why ❔

Fixes API server operation when working on the DB after snapshot recovery.

## Checklist

- [x] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [x] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.
- [x] Code has been formatted via `zk fmt` and `zk lint`.
- [x] Spellcheck has been run via `zk spellcheck`.
- [x] Linkcheck has been run via `zk linkcheck`.

### CODE REVIEW

**Comment by slowli** [comment]
Sorry for combining multiple features into a single PR; it occurred naturally when covering VM-related API methods in tests and debugging related errors. I could try separating this PR into smaller ones (one per each of the 3 items mentioned in "What"?), but I'm not sure it would be worth it.

File: core/lib/zksync_core/src/utils/mod.rs [hunk 1]
```diff
@@ -126,6 +126,31 @@ pub(crate) async fn projected_first_l1_batch(
     Ok(snapshot_recovery.map_or(L1BatchNumber(0), |recovery| recovery.l1_batch_number + 1))
 }
 
+/// Obtains a protocol version projected to be applied for the next miniblock. This is either the version used by the last
+/// sealed miniblock, or (if there are no miniblocks), one referenced in the snapshot recovery record.
+pub(crate) async fn pending_protocol_version(
```
Comment (Deniallugo) [inline core/lib/zksync_core/src/utils/mod.rs:131]: There are two potential problems: 
What if there are no miniblocks yet or on goerli for very ancient blocks there is no protocol version set.

Comment (slowli) [reply core/lib/zksync_core/src/utils/mod.rs:131]: - The case with no miniblocks after snapshot recovery *is* being handled, so this leaves the case without blocks and without snapshot recovery. I don't think we can handle such a case; realistically, it would be most probably caused by `pending_protocol_version()` called "too early" in the node lifecycle. @RomanBrodetski WDYT?
- We load the *latest* miniblock header, so no protocol version for some *ancient* miniblocks is not a concern. The latest miniblock on all nodes should have its protocol version set, right? Even if they don't, it's not a big deal: we could use `ProtocolVersionId::last_potentially_undefined()` as a fallback. This would make `pending_protocol_version()` completely equivalent to the current implementation.

Comment (Deniallugo) [reply core/lib/zksync_core/src/utils/mod.rs:131]: > we could use ProtocolVersionId::last_potentially_undefined() as a fallback

It w'd be great

File: core/lib/zksync_core/src/api_server/execution_sandbox/apply.rs [hunk 2]
```diff
@@ -36,7 +37,250 @@ use super::{
...
+                prev_block_hash: current_l2_block_info.l2_block_hash,
+                // For simplicity we assume each miniblock create one virtual block.
+                // This may be wrong only during transition period.
+                max_virtual_blocks_to_create: 1,
+            }
+        } else if current_l2_block_info.l2_block_number == 0 {
+            // Special case:
+            // - For environments, where genesis block was created before virtual block upgrade it doesn't matter what we put here.
+            // - Otherwise, we need to put actual values here. We cannot create next L2 block with block_number=0 and `max_virtual_blocks_to_create=0`
+            //   because of SystemContext requirements. But, due to intrinsics of SystemContext, block.number still will be resolved to 0.
```
Comment (RomanBrodetski) [inline core/lib/zksync_core/src/api_server/execution_sandbox/apply.rs:138]: given that virtual blocks migration is over, can this be simplified? @slowli good that you extracted this method, but the logic is still a little hard to follow

File: core/lib/zksync_core/src/api_server/web3/mod.rs [hunk 3]
```diff
@@ -468,6 +459,16 @@ impl FullApiParams {
         local_addr_sender: oneshot::Sender<SocketAddr>,
         health_updater: HealthUpdater,
     ) -> anyhow::Result<()> {
+        // We don't want to start the server until at least one L1 batch is stored in Postgres.
+        let earliest_l1_batch_number =
+            wait_for_l1_batch(&self.pool, self.polling_interval, &mut stop_receiver)
+                .await
+                .context("error while waiting for L1 batch in Postgres")?;
```
Comment (RomanBrodetski) [inline core/lib/zksync_core/src/api_server/web3/mod.rs:466]: can we print a log line that we are waiting for it and when it's there? Just to make it easier to debug if we see api server not starting/not accepting requests. If we do this we can also output the block number

File: core/lib/zksync_core/src/utils/mod.rs (hunk 1 above)
Comment (RomanBrodetski) [reply core/lib/zksync_core/src/utils/mod.rs:131]: > so this leaves the case without blocks and without snapshot recovery

So it can only happen if we run state keeper before genesis (or snapshotshot recovery) was complete. I think we need to be more explicit in expectations that component can have, but currently I think the state keeper behaviour is undefined. Now, I can imagine a problem if this method is called from e.g. snapshot recovery logic or genesis logic (as @slowli mentioned, too early in node lifecycle). This shouldn't happen, but it still concerns me that we have methods with undefined behaviour for some DB states.

I guess the easiest would be to call this pre-genesis state (point `1` [from this slack message](https://matter-labs-workspace.slack.com/archives/C05ANUCGCKV/p1707226824450369?thread_ts=1707209111.843459&cid=C05ANUCGCKV)) a special case that the components don't have to handle. This way they can always rely on the fact that there are `miniblocks`/`l1_batches` OR `snapshot_reovery` entires.

**Review by RomanBrodetski — APPROVED** [review]
Looks good, left some very minor comments

File: core/lib/zksync_core/src/utils/mod.rs (hunk 1 above)
Comment (slowli) [reply core/lib/zksync_core/src/utils/mod.rs:131]: > undefined behaviour

Nitpick: this isn't UB in Rust terms, and I think returning an error with intelligible cause is the best possible course of action in such a case, which this logic does.

File: core/lib/zksync_core/src/api_server/execution_sandbox/apply.rs (hunk 2 above)
Comment (slowli) [reply core/lib/zksync_core/src/api_server/execution_sandbox/apply.rs:138]: As you've said, all work on `Sandbox` is essentially just moving code around. I'd prefer to do any semantics changes separately, so it'd be easier to review and revert if necessary.

**Review by RomanBrodetski — APPROVED** [review]
